eclipse/che-theia
# feat(preferences): Allow to define preferences through environment variables

### What does this PR do?
Allow to define extra preferences from environment variables in sidecars (that could come from meta.yaml files)

env variable name is `CHE_THEIA_SIDECAR_PREFERENCES`
env variable value is `stringified value of a JSON content`

### Screenshot/screencast of this PR

https://user-images.githubusercontent.com/436777/112611244-de515500-8e1d-11eb-9dd5-0814636bc3ee.mp4



### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/17975
https://github.com/eclipse/che/issues/19199

### How to test this PR?
Add CHE_THEIA_SIDECAR_PREFERENCES in a meta.yaml env section and look if the preferences is added

here is a devfile example:
```yaml
apiVersion: 1.0.0
metadata:
  name: custom-env
attributes:
  multiRoot: 'on'
components:
  - memoryLimit: 512Mi
    type: cheEditor
    reference: 'https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-1047/che_theia_meta.yaml'
    alias: che-theia
  - type: chePlugin
    reference: 'https://gist.githubusercontent.com/benoitf/ba757696b6591f792487b7a006f21d15/raw/a789b4013053ddbcbd938c3b77a85dd98277d7b9/asciidoctor.yaml'

```

Look at preferences you should see that you've a preference named `asciidoc.helloworld:false`


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=next

Change-Id: I2761540bf307d00319b61e77e3203fe803742c28
Signed-off-by: Florent Benoit <fbenoit@redhat.com>
